V13/type fixes

kmoschcau opened the pull request:
As of 2025-08-03, this needs [this PR's branch](https://github.com/League-of-Foundry-Developers/foundry-vtt-types/pull/3453) as fvtt-types dependency to function.

I encourage reviewers to look at single commits in isolation, instead of the whole diff at the same time. I keep commits atomic for single changes, which should make the whole thing easier to review.

kmoschcau committed:
style: run `npm run format`

kmoschcau committed:
fix(types): add missing IContentPackManifest import in src/global.ts

kmoschcau committed:
fix(types): remove duplicate fromUuidSync declaration

kmoschcau committed:
fix(types): remove reference to non-existent ClientSettings.Values in config

kmoschcau committed:
fix: migrate renderHeadsUpDisplay to renderHeadsUpDisplayContainer hook

kmoschcau committed:
fix(types): widen parameter type of extendTokenConfig to include PrototypeTokenConfig

kmoschcau committed:
fix: add presence filter to LancerToken#getOccupiedSpaces

kmoschcau committed:
fix: itemCreated hook -> createItem

kmoschcau committed:
fix: use sync hook lambda for getSceneNavigationContext

kmoschcau committed:
fix: remove reference to ui.combatCarousel

kmoschcau committed:
fix: drag and drop handling

This expands the types for drag and drop in the system and also adds
more type guards to related functions.

kmoschcau committed:
fix(types): make extendCombatTrackerConfig params more specific

This is one of those cases where foundry violates polymorphism, so we
have to use the more specific application type.

kmoschcau committed:
fix: use explicit data model path in preCreateScene hook

kmoschcau committed:
fix: expose LancerToken hover methods

Also refactor highlightAttackAndDamageTargets to use native HTML. This
might be a breaking change in terms of the use markup. I looked at the
templates and the UUID for the target was directly on the element, the
listener is registered on, so removed the `closest` element query. If
the UUID is not always on that element, that query needs to be put back
in.

LukeAbby commented on src/module/token.ts:
```diff
@@ -114,14 +114,34 @@ export class LancerToken extends Token {
+  /**
+   * This exists only to expose protected hover methods from {@link Token}.
+   */
+  onHoverIn(ev: Canvas.Event.Pointer | MouseEvent): void {
```
An alternative is `LancerToken["onHoverIn"]`; this is basically an escape hatch for the TypeScript error about protected members.

Also this `Canvas.Event.Pointer | MouseEvent`; does this imply a fix is necessary to fvtt-types?

kmoschcau replied:
The MouseEvent is just there to make calling the methods easier. Since they were called from native DOM-Events. The canvas pointer event requires a lot more properties to be added, which are not even used in the two token-methods. `_onHoverOut` doesn't even use the event at all. I'd actually argue we can widen the parameter types in the types for these two methods. Yes, they are documented as being full custom canvas pointer events, but they are not used as such.

LukeAbby commented on src/lancer.ts:
```diff
@@ -384,8 +384,10 @@ Hooks.on("updateActor", (_actor, changes): void => {
-Hooks.on("getSceneNavigationContext", async () => {
-  game.action_manager && (await game.action_manager!.reset());
+Hooks.on("getSceneNavigationContext", () => {
+  // FIXME: This hook has been removed for some time now. If this functionality is still needed, it needs to be placed
+  // on a different hook.
```
I believe `getSceneContextOptions` will do as mentioned on Discord.

LukeAbby commented on src/lancer.ts:
```diff
@@ -457,9 +455,31 @@ Hooks.on("renderCombatTrackerConfig", extendCombatTrackerConfig);
 // Disable token vision and fog exploration by default in scene config
 Hooks.on("preCreateScene", scene => {
-  scene.updateSource({ tokenVision: false, "fog.exploration": false });
+  scene.updateSource({ tokenVision: false, fog: { exploration: false } });
```
Yeah I'd like to add back dotkey support but last time I did it I lagged out repos so bad. Probably some dumb data model stuff.

LukeAbby reviewed (commented):
Obviously a draft PR but I did just want to give a quick review anyways.

BoltsJ reviewed (commented):
Quick review with some requests before you get too far in.

BoltsJ commented:
Also, I find the "conventional commits" style to make it hard for me to read the commit log, and it's not the style used for the repo.

kmoschcau commented on src/module/helpers/dragdrop.ts:
```diff
@@ -260,33 +273,33 @@ export async function resolveNativeDrop(drop: string | FoundryDropData): Promise
       };
     }
-  } else {
+  } else if ("type" in drop) {
```
They might not be necessary at runtime, because you already know that this property will definitely be there. But if our goal is to make the typescript errors go away, these checks are the easier alternative to having to ensure narrower types in more complex ways. This is fairly standard typescript 101 for runtime type checks.

kmoschcau replied:
From my understanding this is a system specific enhancement. It's not part of core.

kmoschcau committed:
Merge remote-tracking branch 'upstream/foundry-v13' into v13/type-fixes

kmoschcau committed:
style(types): change class member access to use no braces

kmoschcau committed:
feat: move the LancerActionManager.reset() call to the ready hook

This was on the getSceneNavigationContext hook before, which has been
removed for some time now. As per suggestion from @Boltsj, this has been
moved to the "ready" hook to figure out if it's still needed.

kmoschcau committed:
fix(types): Mostly revert 78fba57 and rely on widened types in fvtt-types

kmoschcau committed:
feat(types): Improve type safety of SyncUUIDRefField

kmoschcau commented:
@LukeAbby I would like your opinion on https://github.com/Eranziel/foundryvtt-lancer/pull/891/commits/a09ad703a59383de006509d3b1fc3db5f5403878.

kmoschcau committed:
Merge branch 'foundry-v13' into v13/type-fixes

kmoschcau committed:
fix(types): problems in src/module/actor/import.ts

kmoschcau committed:
Merge branch 'foundry-v13' into v13/type-fixes

kmoschcau committed:
fix(types): LancerMECH#system.repairs resolving to `undefined`

There seems to be a bug in `new FullBoundedNumberField()`, which
resolves entire fields created with it to be potentially undefined.
Passing an empty options object to the constructor call fixes this for
some reason.

kmoschcau committed:
fix(types): type errors in lancer-actor.ts

kmoschcau committed:
docs: add a TODO-comment concerning FullBoundedNumberField replacement

kmoschcau committed:
fix(types): type errors in loadout-util.ts

kmoschcau committed:
fix(types): most errors in mech-sheet.ts

kmoschcau committed:
fix(types): type errors in npc-sheet.ts

kmoschcau committed:
fix(types): type errors in pilot-sheet.ts

kmoschcau committed:
fix(types): most type errors in acc_diff plugin

kmoschcau committed:
Merge branch 'foundry-v13' into v13/type-fixes

kmoschcau committed:
fix(types): errors in acc_diff/spotter.ts

kmoschcau committed:
fix(types): simplify some fixes and add more HACK comments

kmoschcau committed:
fix(types): remove unused import in loadout-util.ts

kmoschcau committed:
style: remove unintended formatting changes

kmoschcau committed:
docs(types): add TODO comment to import.ts

kmoschcau committed:
fix(types): make pilot level non-nullable

kmoschcau committed:
fix(types): make npc tier non-nullable

kmoschcau committed:
fix(types): damage HUD plugin

kmoschcau committed:
fix(types): some other misc type complaints

kmoschcau committed:
fix(types): some more errors in import

kmoschcau committed:
fix(types): errors in lancer-actor.ts

kmoschcau committed:
fix(types): errors in loadout-util.ts

kmoschcau committed:
fix(types): declare object type in LancerActorSheet

kmoschcau committed:
fix(types): declare object type in LancerItemSheet

kmoschcau committed:
fix(types): errors in mech-sheet.ts

kmoschcau committed:
fix(types): errors in lancer-initiative-config-form.ts

kmoschcau committed:
fix(types): errors in weapon-range-template.ts

kmoschcau committed:
fix(types): errors in lancer-combat-tracker.ts

kmoschcau committed:
fix(types): errors around lancer-combat.ts

kmoschcau committed:
fix(types): error in comp-builder.ts

kmoschcau committed:
fix: type errors in config.ts

kmoschcau committed:
fix(types): errors around converter.ts

kmoschcau committed:
fix(types): errors around effector.ts

kmoschcau committed:
fix(types): errors in lancer-active-effect.ts

kmoschcau committed:
fix(types): errors in flows/_template.ts

kmoschcau committed:
fix(types): errors in flows/activation.ts

kmoschcau committed:
fix(types): errors in flows/attack.ts

kmoschcau committed:
fix(types): errors in bits/power.ts

kmoschcau committed:
feat(types): add a dice-so-nice stub

kmoschcau committed:
fix(types): errors in flows/cascade.ts

Also unify HACK:-comments

kmoschcau committed:
fix(types): errors around flows/damage.ts

kmoschcau committed:
fix(types): flow hooks

kmoschcau committed:
fix(types): add RollTerm.Options.rollOrder for dsn

kmoschcau committed:
refactor: correct naming of Flow hook caller methods

kmoschcau committed:
fix(types): add Flow#steps getter and overrides

kmoschcau committed:
Merge branch 'master' into v13/type-fixes

kmoschcau committed:
fix(types): remove unused @ts-expect-error

kmoschcau committed:
fix(types): errors in flows/item.ts

kmoschcau committed:
fix(types): errors in flows/npc.ts

kmoschcau committed:
fix(types): errors in flows/overheat.ts

kmoschcau committed:
Merge branch 'master' into v13/type-fixes

# Conflicts:
#	src/module/actor/lancer-actor-sheet.ts
#	src/module/item/item-sheet.ts

kmoschcau committed:
refactor(types): move active effect hook declaration closer to use site

kmoschcau committed:
fix(types): error in register-flows.ts

kmoschcau committed:
fix(types): error in flows/structure.ts

kmoschcau committed:
fix: make action model "name" required

kmoschcau committed:
fix(types): error in helpers/commons.ts

kmoschcau committed:
fix(types): remove "any" type annotation from "fields" in shared.ts

kmoschcau committed:
fix: handleNPCExport function

kmoschcau committed:
refactor: remove helpers/io.ts

The functions in there are not used and are full of errors.

kmoschcau committed:
fix(types): errors in pilot-sheet.ts

kmoschcau committed:
fix: type errors around helpers/item.ts

kmoschcau committed:
fix: type errors around helpers/loadout.ts

kmoschcau committed:
fix(types): missing default options for most custom SchemaFields

kmoschcau committed:
fix: type errors in lancer-item.ts

kmoschcau committed:
fix(types): custom tour step properties

kmoschcau committed:
fix: wrong InitializedType for DamageField

kmoschcau committed:
fix(types): Tour.Step decl merge conflict

kmoschcau committed:
fix(types): wide effectBox text parameter type

kmoschcau committed:
fix(types): too strict icon map in text-enrichers.ts

kmoschcau committed:
fix(types): error in actor/import.ts

kmoschcau committed:
fix: type errors in lancer-actor.ts

kmoschcau committed:
fix(types): errors in loadout-util.ts

kmoschcau committed:
fix: type errors in mech-sheet.ts

kmoschcau committed:
fix: type errors in npc-sheet.ts

kmoschcau committed:
fix: type errors in effector.ts

kmoschcau committed:
fix: type errors in item-utils.ts

kmoschcau committed:
fix: type errors in from-lid.ts

kmoschcau committed:
fix(types): error in helpers/hooks.ts

kmoschcau committed:
fix: type errors in flows/_render.ts

kmoschcau committed:
fix(types): add missing return type annotation in bits/synergy.ts
